batch/batch/driver/pool.py

```diff
@@ -0,0 +1,469 @@
+  standing_worker = %s, standing_worker_cores = %s,
+  max_instances = %s, max_live_instances = %s
+WHERE name = %s;
+''',
+            (worker_cores, worker_disk_size_gb,
+             worker_local_ssd_data_disk, worker_pd_ssd_data_disk_size_gb,
+             standing_worker, standing_worker_cores,
+             max_instances, max_live_instances, self.name))
+
+        self.worker_cores = worker_cores
+        self.worker_disk_size_gb = worker_disk_size_gb
+        self.worker_local_ssd_data_disk = worker_local_ssd_data_disk
+        self.worker_pd_ssd_data_disk_size_gb = worker_pd_ssd_data_disk_size_gb
+        self.standing_worker = standing_worker
+        self.standing_worker_cores = standing_worker_cores
+        self.max_instances = max_instances
+        self.max_live_instances = max_live_instances
+
+    def adjust_for_remove_instance(self, instance):
+        super().adjust_for_remove_instance(instance)
+        if instance in self.healthy_instances_by_free_cores:
+            self.healthy_instances_by_free_cores.remove(instance)
+
+    def adjust_for_add_instance(self, instance):
+        super().adjust_for_add_instance(instance)
+        if (instance.state == 'active'
+                and instance.failed_request_count <= 1):
+            self.healthy_instances_by_free_cores.add(instance)
+
+    async def create_instance(self, cores=None, max_idle_time_msecs=None):
+        if cores is None:
+            cores = self.worker_cores
+
+        if max_idle_time_msecs is None:
+            max_idle_time_msecs = WORKER_MAX_IDLE_TIME_MSECS
+
+        machine_name = self.generate_machine_name()
+
+        zone = self.zone_monitor.get_zone(cores, self.worker_local_ssd_data_disk,
+                                          self.worker_pd_ssd_data_disk_size_gb)
+        if zone is None:
+            return
+
+        activation_token = secrets.token_urlsafe(32)
+        instance = await Instance.create(self.app, self, machine_name, activation_token, cores * 1000, zone)
+        self.add_instance(instance)
+        log.info(f'created {instance}')
+
+        machine_type = f'n1-{self.worker_type}-{cores}'
+        await create_instance(app=self.app, zone=zone, machine_name=machine_name,
+                              machine_type=machine_type, activation_token=activation_token,
+                              max_idle_time_msecs=max_idle_time_msecs,
+                              worker_local_ssd_data_disk=self.worker_local_ssd_data_disk,
+                              worker_pd_ssd_data_disk_size_gb=self.worker_pd_ssd_data_disk_size_gb,
+                              worker_disk_size_gb=self.worker_disk_size_gb)
+
+    async def control_loop(self):
+        log.info(f'starting control loop for {self}')
+        while True:
+            try:
+                ready_cores = await self.db.select_and_fetchone(
+                    '''
+SELECT CAST(COALESCE(SUM(ready_cores_mcpu), 0) AS SIGNED) AS ready_cores_mcpu
```

Comment:
We should not need the cast or coalesce. ready_cores_mcpu is a not null.

Reply:
I disagree and am going to strongly push back here! If there are no rows that match, then SUM() will return NULL. The SUM() doesn't return the right type which is why we had to CAST it. We can try looking for the original PRs where these were added.

Reply:
Ugh, this grinds my gears.